### PR TITLE
QuizSessionViewController

### PR DESCRIPTION
**Pull request by barbarakos**
Added:
- QuestionView
- QuizSessionViewModel
- QuizSessionViewController

Still need to add:
- Coloring of progress bar views when the question is answered
- Switching to new question

Simulator Screen Shot - iPhone 14 Plus - 2022-11-08 at 11 24 25

### CODE REVIEW

File: QuizApp/Quiz/ViewControllers/QuizDetailsViewController.swift [hunk 1]
```diff
@@ -65,7 +69,12 @@ extension QuizDetailsViewController: ConstructViewsProtocol {
         view.addSubview(leaderboardButton)
 
         quizDetailsView = QuizDetailsView()
-        quizDetailsView.set(for: quizDetailsViewModel.quiz)
+        quizDetailsView
+            .startButtonTapped
+            .sink { _ in
+                self.startQuiz()
+            }
+            .store(in: &cancellables)
```
Comment (dorastipkovic) [inline QuizApp/Quiz/ViewControllers/QuizDetailsViewController.swift]: Move binding to a separate func.

File: QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        buildViews()
+        bindViewModel()
+    }
+
+    private func setQuestionView(question: QuestionModel) {
+        setProgressViewColor()
+        questionNumberLabel.text = "\(currentQuestionNum)/\(viewModel.quiz.numberOfQuestions)"
+        questionView.setQuestion(question: question)
+    }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift]: private functions after public

File: QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    func nextQuestion() {
+        currentQuestionNum+=1
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift]:
```suggestion
        currentQuestionNum += 1
```

File: QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        setQuestionView(question: question)
+    }
+
+    func bindViewModel() {
+        viewModel
+            .$questions
+            .sink { [weak self] questions in
+                if !questions.isEmpty {
+                    self?.setQuestionView(question: questions[0])
+                }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift]: change to guard syntax

File: QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            $0.leading.trailing.bottom.equalToSuperview().inset(questionInsets)
+        }
+    }
+
+}
+
+// MARK: ProgressBar functions
+extension QuizSessionViewController {
+
+    func setProgressViewColor() {
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewControllers/QuizSessionViewController.swift]: Can this also be private?

File: QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        self.useCase = quizSessionUseCase
+        self.quiz = quiz
+        fetchQuestions()
+    }
+
+    func fetchQuestions() {
+        Task {
+            do {
+                let quizSession = QuizSessionModel(from: try await useCase.fetchQuestions(quizId: quiz.id))
+                DispatchQueue.main.async { [weak self] in
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift:25]: newline before

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+
+    private func colorCorrectAnswer() {
+        stackView
+            .subviews
+            .forEach { button in
+                guard let button = button as? Button, button.id == correctAnswerId else {
+                    return
+                }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: return in same line if fits

File: QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import Combine
+import UIKit
+
+class QuizSessionViewModel {
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift:4]: Instead of storing currentQuestionIndex, viewModel could publish a current question model. 
That way when you call viewModel.nextQuestion() new question is published and viewcontroller would be subscribed to it.

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        buildViews()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setQuestion(question: QuestionModel) {
+        self.answers = question.answers
+        self.correctAnswerId = question.correctAnswerId
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: self not needed

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                .sink { [weak self] _ in
+                    guard let self = self else { return }
+
+                    let isCorrect = answer.id == self.correctAnswerId
+                    answerButton.backgroundColor = isCorrect ? .green : .red
+
+                    if !isCorrect {
+                        self.colorCorrectAnswer()
+                    }
+
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: Why is this needed when you have `answerButton.backgroundColor = isCorrect ? .green : .red` ?

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/QuizSession/Views/QuestionView.swift (hunk 10 above)
Comment (barbarakos) [reply QuizApp/QuizSession/Views/QuestionView.swift]: In case the chosen answer is not correct, with ```answerButton.backgroundColor = isCorrect ? .green : .red``` it's color changes to red, but then I have to go through all the answers and color the correct one green

Comment (dorastipkovic) [reply QuizApp/QuizSession/Views/QuestionView.swift]: That logic is a bit confusing, and not obvious. 
Instead of having both `answerButton.backgroundColor = isCorrect ? .green : .red ` and `colorCorrectAnswer()`, keep only the function but then go through all buttons and color them in the appropriate color.

File: QuizApp/Extensions/UIButton+Extensions.swift [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    override var intrinsicContentSize: CGSize {
+            let labelSize = titleLabel?.sizeThatFits(CGSize(
+                                        width: frame.size.width,
+                                        height: CGFloat.greatestFiniteMagnitude)) ?? .zero
+            let desiredButtonSize = CGSize(width: labelSize.width + 30, height: labelSize.height + 40)
+
+            return desiredButtonSize
+        }
```
Comment (dorastipkovic) [inline QuizApp/Extensions/UIButton+Extensions.swift]: Why do you override intrinsicContentSize instead of using autolayout constraints?

File: QuizApp/Extensions/UIButton+Extensions.swift [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import UIKit
+
+class Button: UIButton {
```
Comment (dorastipkovic) [inline QuizApp/Extensions/UIButton+Extensions.swift]: Rename to `IdentifiableButton` or smth that will make the purpose clear

File: QuizApp/Quiz/ViewControllers/QuizDetailsViewController.swift [hunk 13]
```diff
@@ -65,7 +69,12 @@ extension QuizDetailsViewController: ConstructViewsProtocol {
         view.addSubview(leaderboardButton)
 
         quizDetailsView = QuizDetailsView()
-        quizDetailsView.set(for: quizDetailsViewModel.quiz)
+        quizDetailsView
+            .startButtonTapped
+            .sink { _ in
```
Comment (dorastipkovic) [inline QuizApp/Quiz/ViewControllers/QuizDetailsViewController.swift]: weak self

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            let title = NSAttributedString(
+                string: answer.answer,
+                attributes: [.font: UIFont.systemFont(ofSize: 20, weight: .semibold),
+                             .foregroundColor: UIColor.white]
+            )
+            answerButton.titleLabel?.numberOfLines = 0
+            answerButton.titleLabel?.lineBreakMode = .byWordWrapping
+            answerButton.setAttributedTitle(title, for: .normal)
+            answerButton.contentHorizontalAlignment = UIControl.ContentHorizontalAlignment.leading
+            answerButton.addConstraint(
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: use snapkit

File: QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Published var currentQuestion: QuestionModel!
+    var quiz: QuizModel
+    var questions: [QuestionModel] = []
+    var currentQuestionIndex: Int = 0
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewModel/QuizSessionViewModel.swift]: You are tracking currentQuestionIndex in two places (vc and model), which is dangerous cause it can easily lead to inconsistent states. Figure out how you can avoid using a counter for currentQuestionIndex.

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+                }
+                .store(in: &cancellables)
+
+            stackView.addArrangedSubview(answerButton)
+        }
+    }
+
+    private func colorAnswers(selectedAnswer: IdentifiableButton, isCorrect: Bool) {
+        selectedAnswer.backgroundColor = isCorrect ? .green : .red
+        if !isCorrect {
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: change to guard

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+            stackView
+                .subviews
+                .forEach { button in
+                    guard let button = button as? IdentifiableButton, button.id == correctAnswerId else { return }
+
+                    button.backgroundColor = .green
+                }
+        }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: try like this
```suggestion
            stackView
                .subviews
                .compactMap { $0 as? IdentifiableButton }
                .filter { $0.id == correctAnswerId }
                .forEach { $0.backgroundColor = .green }
        }
```

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/QuizSession/ViewModel/Models/QuizSessionModel.swift [hunk 18]
```diff
@@ -5,7 +5,7 @@ struct QuizSessionModel {
 
     init(from model: QuizSessionUseCaseModel) {
         sessionId = model.sessionId
-        questions = model.questions.map { QuestionModel(from: $0) }
+        questions = model.questions.map { QuestionModel(from: $0, index: model.questions.firstIndex(of: $0)!) }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewModel/Models/QuizSessionModel.swift]: use enumerated() to get array indexes

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 19]
```diff
@@ -48,15 +48,9 @@ class QuestionView: UIView {
...
-                NSLayoutConstraint(
-                    item: answerButton,
-                    attribute: .height,
-                    relatedBy: .equal,
-                    toItem: answerButton.titleLabel,
-                    attribute: .height,
-                    multiplier: 1.0,
-                    constant: 40))
+            answerButton.snp.makeConstraints {
+                $0.height.equalTo(answerButton.titleLabel!.snp.height).offset(40)
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: guard answerButton.titleLabel instead of force unwrapping

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/QuizSession/Views/QuestionView.swift [hunk 20]
```diff
@@ -44,8 +44,13 @@ class QuestionView: UIView {
                 attributes: [.font: UIFont.systemFont(ofSize: 20, weight: .semibold),
                              .foregroundColor: UIColor.white]
             )
-            answerButton.titleLabel?.numberOfLines = 0
-            answerButton.titleLabel?.lineBreakMode = .byWordWrapping
+
+            guard let answerTitleLabel = answerButton.titleLabel else {
+                return
+            }
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/Views/QuestionView.swift]: fits in one line
you're still force unwrapping below: `answerButton.titleLabel!.snp.height`

File: QuizApp/QuizSession/ViewModel/Models/QuizSessionModel.swift [hunk 21]
```diff
@@ -5,7 +5,12 @@ struct QuizSessionModel {
 
     init(from model: QuizSessionUseCaseModel) {
         sessionId = model.sessionId
-        questions = model.questions.map { QuestionModel(from: $0, index: model.questions.firstIndex(of: $0)!) }
+        questions = model.questions.map { (ucModel) -> QuestionModel in
+            QuestionModel(
+                from: ucModel,
+                index: model.questions.enumerated()
+                    .first(where: { $0.element == ucModel})
+                    .map { $0.offset }!)}
```
Comment (dorastipkovic) [inline QuizApp/QuizSession/ViewModel/Models/QuizSessionModel.swift]:
```
        questions = model.questions
            .enumerated()
            .map { index, model in
            QuestionModel(from: model, index: index)
        }
```

**Review by dorastipkovic — APPROVED** [review]
